controllers/grafananotificationtemplate_controller.go

```diff
@@ -0,0 +1,260 @@
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+
+	defer func() {
+		if err := r.Client.Status().Update(ctx, notificationTemplate); err != nil {
+			r.Log.Error(err, "updating status")
+		}
+		if meta.IsStatusConditionTrue(notificationTemplate.Status.Conditions, conditionNoMatchingInstance) {
+			controllerutil.RemoveFinalizer(notificationTemplate, grafanaFinalizer)
+		} else {
+			controllerutil.AddFinalizer(notificationTemplate, grafanaFinalizer)
+		}
+		if err := r.Update(ctx, notificationTemplate); err != nil {
+			r.Log.Error(err, "failed to set finalizer")
+		}
+	}()
+
+	instances, err := r.GetMatchingInstances(ctx, notificationTemplate, r.Client)
+	if err != nil {
+		setNoMatchingInstance(&notificationTemplate.Status.Conditions, notificationTemplate.Generation, "ErrFetchingInstances", fmt.Sprintf("error occurred during fetching of instances: %s", err.Error()))
+		meta.RemoveStatusCondition(&notificationTemplate.Status.Conditions, conditionNotificationTemplateSynchronized)
+		r.Log.Error(err, "could not find matching instances")
+		return ctrl.Result{RequeueAfter: RequeueDelay}, err
+	}
+
+	if len(instances) == 0 {
+		meta.RemoveStatusCondition(&notificationTemplate.Status.Conditions, conditionNotificationTemplateSynchronized)
+		setNoMatchingInstance(&notificationTemplate.Status.Conditions, notificationTemplate.Generation, "EmptyAPIReply", "Instances could not be fetched, reconciliation will be retried")
+		return ctrl.Result{}, nil
+	}
+
+	removeNoMatchingInstance(&notificationTemplate.Status.Conditions)
+
+	applyErrors := make(map[string]string)
+	for _, grafana := range instances {
+		// can be removed in go 1.22+
+		grafana := grafana
+		if grafana.Status.Stage != grafanav1beta1.OperatorStageComplete || grafana.Status.StageStatus != grafanav1beta1.OperatorStageResultSuccess {
+			controllerLog.Info("grafana instance not ready", "grafana", grafana.Name)
+			continue
+		}
+
+		err := r.reconcileWithInstance(ctx, &grafana, notificationTemplate)
+		if err != nil {
+			applyErrors[fmt.Sprintf("%s/%s", grafana.Namespace, grafana.Name)] = err.Error()
+		}
+	}
+	condition := metav1.Condition{
+		Type:               conditionNotificationTemplateSynchronized,
+		ObservedGeneration: notificationTemplate.Generation,
+		LastTransitionTime: metav1.Time{
+			Time: time.Now(),
+		},
+	}
+
+	if len(applyErrors) == 0 {
+		condition.Status = metav1.ConditionTrue
+		condition.Reason = conditionApplySuccessful
+		condition.Message = fmt.Sprintf("Notification template was successfully applied to %d instances", len(instances))
+	} else {
+		condition.Status = metav1.ConditionFalse
+		condition.Reason = conditionApplyFailed
+
+		var sb strings.Builder
+		for i, err := range applyErrors {
+			sb.WriteString(fmt.Sprintf("\n- %s: %s", i, err))
+		}
+
+		condition.Message = fmt.Sprintf("Notification template failed to be applied for %d out of %d instances. Errors:%s", len(applyErrors), len(instances), sb.String())
+	}
```

Comment:
There's a handy `buildSynchronizedCondition` function for building the final condition!
```suggestion
	condition := buildSynchronizedCondition("Folder", conditionNotificationTemplateSynchronized, notificationTemplate.Generation, applyErrors, len(instances))
	meta.SetStatusCondition(&notificationTemplate.Status.Conditions, condition)

	if len(applyErrors) > 0 {
		return ctrl.Result{RequeueAfter: RequeueDelay}, fmt.Errorf("failed to apply to all instances: %v", applyErrors)
	}
```